### DISCUSSION TITLE
Per Pixel setting performance on Retroid Pocket 5 (SD865 w/ Android 13)

### DISCUSSION THREAD

**Discussion by banana3333** [flyinghead/flycast#1765] [General]
The GPU can actually properly support the setting unlike most of these handhelds with Mali GPUs, which is awesome, but the performance is terrible. Even at 1x resolution with the RP5 system setting on "High Performance", I'm still running into serious dips and slowdowns. Is this just how it is for this setting with the amount of power the chip gives? I understand it's ultimately a very demanding setting.

Possibly related: I'm playing Bangai-O with the Per Triangle setting, 1x resolution, the RP5 system setting on "High Performance", and I'm running into similarly terrible performance. I've tried OpenGL and Vulkan, lowering and outright disabling Anisotropic Filtering, unchecking settings like "Delay Frame Swapping" and "Fix Upscale Bleeding Edge", forcing HLE BIOS, etc and none of it seems to make a difference. I'm pretty sure something is up with that, at least. Qualcomm had only just gotten around to helping Retroid get Android 13 working on this chipset in the last two months or so, so I'm wondering if this is an issue born out of that somehow.

![Screenshot_20241202-144940](https://github.com/user-attachments/assets/5d7da861-106f-441f-9355-0e74a7de568e)

This is with the 11-22-2024 build of Flycast, if that makes any difference.

**Comment by flyinghead** [comment]
I believe this hardware is much more capable than that. Here is a POCO F1 (Adreno 630) running with Vulkan per pixel at x1 resolution:
![image](https://github.com/user-attachments/assets/ccef9e53-ec08-4866-ade9-9e10f56f1386)

Adreno drivers are notoriously a hit or miss. In this case it's not that bad since it's kinda working. (I've seen higher end adreno GPUs with completely broken drivers only showing a black screen.) But the performance isn't what is expected obviously.
You could try custom adreno vulkan drivers (See Settings > About when vulkan is selected) and you should get better performance even with the open source turnip driver. Proprietary Qualcomm drivers will give you better performance if you're lucky to find one that work at all...

**Comment by banana3333** [comment]
Oh dang, I had no idea about that custom driver option! I tried it with the Turnip driver and the difference is absolutely night and day. I can play everything I've tried at 1x with the Per Pixel setting without issue, and many at 1.5x. Bangai-O is still running the same, which is strange. Not super fussed about, though.

**Reply by flyinghead** [reply]
Great news!
The issue with Bangai-O is probably a different one since it's not happening in per pixel and this game doesn't normally require a fast GPU.

**Reply by dantesquivel** [reply]
> Oh dang, I had no idea about that custom driver option! I tried it with the Turnip driver and the difference is absolutely night and day. I can play everything I've tried at 1x with the Per Pixel setting without issue, and many at 1.5x. Bangai-O is still running the same, which is strange. Not super fussed about, though.

Which Turnip driver did you use that worked for you?
